src/3d/qgsrulebasedchunkloader_p.cpp

```diff
@@ -203,4 +205,58 @@ void QgsRuleBasedChunkedEntity::onTerrainElevationOffsetChanged( float newOffset
   mTransform->setTranslation( QVector3D( 0.0f, newOffset, 0.0f ) );
 }
 
+QVector<QgsRayCastingUtils::RayHit> QgsRuleBasedChunkedEntity::rayIntersection( const QgsRayCastingUtils::Ray3D &ray, const QgsRayCastingUtils::RayCastContext &context ) const
+{
+  Q_UNUSED( context )
+  QgsDebugMsgLevel( QStringLiteral( "Ray cast on vector layer" ), 2 );
+  int nodeUsed = 0;
+  int nodesAll = 0;
+  int hits = 0;
+  QVector<QgsRayCastingUtils::RayHit> result;
+
+  float minDist = -1;
+  QVector3D intersectionPoint;
+  QgsFeatureId fid;
+  QgsFeatureId nearestFid;
+
+  const QList<QgsChunkNode *> activeNodes = this->activeNodes();
+  for ( QgsChunkNode *node : activeNodes )
+  {
+    nodesAll++;
+    if ( node->entity() &&
+         ( minDist < 0 || node->bbox().distanceFromPoint( ray.origin() ) < minDist ) &&
+         QgsRayCastingUtils::rayBoxIntersection( ray, node->bbox() ) )
+    {
+      nodeUsed++;
+      const QList<Qt3DRender::QGeometryRenderer *> rendLst = node->entity()->findChildren<Qt3DRender::QGeometryRenderer *>();
+      for ( const auto &rend : rendLst )
+      {
+        auto *geom = rend->geometry();
+        QgsTessellatedPolygonGeometry *polygonGeom = qobject_cast<QgsTessellatedPolygonGeometry *>( geom );
+        if ( !polygonGeom )
+          continue; // other QGeometry types are not supported for now
+
+        QVector3D nodeIntPoint;
+        if ( polygonGeom->rayIntersection( ray, mTransform->matrix(), nodeIntPoint, fid ) )
+        {
+          hits++;
+          float dist = ( ray.origin() - nodeIntPoint ).length();
+          if ( minDist < 0 || dist < minDist )
+          {
+            minDist = dist;
+            intersectionPoint = nodeIntPoint;
+            nearestFid = fid;
+          }
+        }
+      }
+    }
+  }
+  if ( !intersectionPoint.isNull() )
+  {
+    QgsRayCastingUtils::RayHit hit( minDist, intersectionPoint, nearestFid );
+    result.append( hit );
+  }
+  QgsDebugMsgLevel( QStringLiteral( "Active Nodes: %1, checked nodes: %2, hits found: %3" ).arg( nodesAll ).arg( nodeUsed ).arg( hits ), 2 );
+  return result;
+}
```

Comment:
This looks like a verbatim copy of what is in `QgsVectorLayerChunkedEntity::rayIntersection()` - I would suggest to have maybe the QgsVectorLayerChunkedEntity ray intersection implementation available also as a static method in QgsVectorLayerChunkedEntity, so that the rule-based implementation could call it

Reply:
fixed